### DISCUSSION TITLE
What kind of queries are not working with Avram?

### DISCUSSION THREAD

**Discussion by paulcsmith** [luckyframework/lucky#1153] [General]
SQL can get pretty wild! Sometimes Avram can’t get you what you want, but we’d like to be able to handle most queries out of the box. 
 
Please post here with any queries that require custom SQL so we can try to get them working :D

**Comment by paulcsmith** [comment]
From @confact 

```

def get_pages
    sql = <<-SQL
    SELECT path as address, COUNT(id) as count FROM events
    WHERE domain_id=#{@domain.id} AND created_at > '#{period_days}'
    GROUP BY path
    ORDER BY COUNT(id) desc LIMIT 10;
    SQL
    pages = AppDatabase.run do |db|
      db.query_all sql, as: StatsPages
    end
    pages = count_percentage(pages)
    return pages
  end

```

**Comment by paulcsmith** [comment]
Bonus points if you also include an example in another library where the query would work. That way we can see how others handle it. Thank you!

**Comment by jwoertink** [comment]
```
SELECT name FROM users;
```

```ruby
# from rails
User.select(:name)
```

**Reply by paulcsmith** [reply]
Is this done for performance reasons? I typically do `UserQuery.new.map(&.name)` but maybe that doesn't work for all use-cases? Would love to learn more about this

**Reply by jwoertink** [reply]
Yeah, for performance reasons, but also in the case of dynamic columns which becomes a whole other topic. For example: In rails, you can do

```ruby
user = User.select("COUNT(friends) as how_many_friends_do_i_have").first
user.how_many_friends_do_i_have #=> 2
```

I don't see how we would be able to implement this in a type safe way, but I figured it should at least be documented in case one of us has that "lightbulb" moment some day 😅

**Reply by jkthorne** [reply]
would `UserQuery` query the entire row and then select name.  I think the Rails version only returns the field `name` not the whole row.

**Comment by confact** [comment]
Oh, one more:
```SQL
SELECT COUNT(DISTINCT user_id) FROM events WHERE domain_id=#{@domain.id} AND created_at > '#{period_days}';
```
```ruby
# from rails:
Event.where("created_at > ?", period_days).where(domain_id: @domain.id).distinct.count(:user_id)
```

**Reply by paulcsmith** [reply]
@confact I *think* this would work with this:

```crystal
period_days = 5.days.ago # or minutes, seconds, whatever
EventQuery.new.created_at.gt(period_days).domain_id(@domain.id).distinct_on(&.user_id).select_count
```

Does that work? If not can you let me know what the issue is or open an issue? Because this should be working now

**Comment by matthewmcgarvey** [comment]
Queries using joins can produce invalid sql https://github.com/luckyframework/avram/issues/370

**Reply by paulcsmith** [reply]
Ruh-roh! Well we definitely want to fix that :D
